Mathlib/NumberTheory/NumberField/Units.lean

```diff
@@ -149,4 +165,405 @@ theorem rootsOfUnity_eq_torsion [NumberField K] :
+    replace hyp := (le_of_eq hyp).trans (norm_sum_le _ _)
+    simp_rw [norm_mul, norm_neg, Real.norm_eq_abs, Nat.abs_cast] at hyp
+    refine (le_trans ?_ hyp).trans ?_
+    · rw [← hw]
+      exact tool _ (abs_nonneg _)
+    · refine (Finset.sum_le_card_nsmul Finset.univ _  _
+        (fun w _ => logEmbedding_component_le hr h w)).trans ?_
+      rw [nsmul_eq_mul]
+      refine mul_le_mul ?_ le_rfl hr (Fintype.card (InfinitePlace K)).cast_nonneg
+      simp [Finset.card_univ]
+  · have hyp := logEmbedding_component_le hr h ⟨w, hw⟩
+    rw [logEmbedding_component, abs_mul, Nat.abs_cast] at hyp
+    refine (le_trans ?_ hyp).trans ?_
+    · exact tool _ (abs_nonneg _)
+    · nth_rw 1 [← one_mul r]
+      exact mul_le_mul (Nat.one_le_cast.mpr Fintype.card_pos) (le_of_eq rfl) hr (Nat.cast_nonneg _)
+
+variable (K)
+
+/-- The lattice formed by the image of the logarithmic embedding. -/
+noncomputable def _root_.NumberField.Units.unitLattice :
+    AddSubgroup ({w : InfinitePlace K // w ≠ w₀} → ℝ) :=
+  AddSubgroup.map (logEmbedding K) ⊤
+
+theorem unitLattice_inter_ball_finite (r : ℝ) :
+    ((unitLattice K : Set ({ w : InfinitePlace K // w ≠ w₀} → ℝ)) ∩
+      Metric.closedBall 0 r).Finite := by
+  obtain hr | hr := lt_or_le r 0
+  · convert Set.finite_empty
+    rw [Metric.closedBall_eq_empty.mpr hr]
+    exact Set.inter_empty _
+  · suffices Set.Finite {x : (𝓞 K)ˣ | IsIntegral ℤ (x : K) ∧
+        ∀ (φ : K →+* ℂ), ‖φ x‖ ≤ Real.exp ((Fintype.card (InfinitePlace K)) * r)} by
+      refine (Set.Finite.image (logEmbedding K) this).subset ?_
+      rintro _ ⟨⟨x, ⟨_, rfl⟩⟩, hx⟩
+      refine ⟨x, ⟨x.val.prop, (le_iff_le _ _).mp (fun w => (Real.log_le_iff_le_exp ?_).mp ?_)⟩, rfl⟩
+      · exact pos_iff.mpr (coe_ne_zero x)
+      · rw [mem_closedBall_zero_iff] at hx
+        exact (le_abs_self _).trans (log_le_of_logEmbedding_le hr hx w)
+    refine Set.Finite.of_finite_image ?_ ((coe_injective K).injOn _)
+    refine (Embeddings.finite_of_norm_le K ℂ
+        (Real.exp ((Fintype.card (InfinitePlace K)) * r))).subset ?_
+    rintro _ ⟨x, ⟨⟨h_int, h_le⟩, rfl⟩⟩
+    exact ⟨h_int, h_le⟩
+
+section span_top
+
+-- To prove that the span over `ℝ` of the `unitLattice` is equal to the full space, we construct
+-- for each infinite place `w₁ ≠ w₀` a unit `u_w₁` of `K` such that, for all infinite places
+-- `w` such that `w ≠ w₁`, we have `Real.log w (u_w₁) < 0` (and thus `Real.log w₁ (u_w₁) > 0`).
+-- It follows then from a determinant computation (using `Matrix.det_ne_zero_of_neg`) that the
+-- image by `logEmbedding` of these units is a `ℝ`-linearly independent family.
+-- The unit `u_w₁` is obtained by construction a sequence `seq n` of nonzero algebraic integers
+-- that is strictly decreasing at infinite places distinct from `w₁` and of norm `≤ B`. Since
+-- there are finitely many ideals of norm `≤ B`, there exists two terms in the sequence defining
+-- the same ideal and their quotient is the desired unit `u_w₁` (see `exists_unit`).
```

Comment:
```suggestion
-- To prove that the span over `ℝ` of the `unitLattice` is equal to the full space, we construct
-- for each infinite place `w₁ ≠ w₀` a unit `u_w₁` of `K` such that, for all infinite places
-- `w` such that `w ≠ w₁`, we have `Real.log w (u_w₁) < 0` (and thus `Real.log w₁ (u_w₁) > 0`).
-- It follows then from a determinant computation (using `Matrix.det_ne_zero_of_neg`) that the
-- image by `logEmbedding` of these units is a `ℝ`-linearly independent family.
-- The unit `u_w₁` is obtained by constructing a sequence `seq n` of nonzero algebraic integers
-- that is strictly decreasing at infinite places distinct from `w₁` and of norm `≤ B`. Since
-- there are finitely many ideals of norm `≤ B`, there exists two term in the sequence defining
-- the same ideal and their quotient is the desired unit `u_w₁` (see `exists_unit`).
```
Again you can make this visible in the doc.